common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -1018,51 +1019,48 @@ private static boolean isRoot0() {
         return false;
     }
 
-    @SuppressWarnings("LoopStatementThatDoesntLoop")
     private static int javaVersion0() {
-        int javaVersion;
+        final int majorVersion;
 
-        // Not really a loop
-        for (;;) {
-            // Android
-            if (isAndroid()) {
-                javaVersion = 6;
-                break;
-            }
+        if (isAndroid()) {
+            majorVersion = 6;
+        } else {
+            majorVersion = majorVersionFromJavaSpecificationVersion();
+        }
 
-            try {
-                Method getVersion = java.lang.Runtime.class.getMethod("version");
-                Object version = getVersion.invoke(null);
-                javaVersion = (Integer) version.getClass().getMethod("major").invoke(version);
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        logger.debug("Java version: {}", majorVersion);
 
-            try {
-                Class.forName("java.time.Clock", false, getClassLoader(Object.class));
-                javaVersion = 8;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        return majorVersion;
+    }
 
-            try {
-                Class.forName("java.util.concurrent.LinkedTransferQueue", false, getClassLoader(BlockingQueue.class));
-                javaVersion = 7;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+    static int majorVersionFromJavaSpecificationVersion() {
+        try {
+            final String javaSpecVersion = AccessController.doPrivileged(new PrivilegedAction<String>() {
+                @Override
+                public String run() {
+                    return System.getProperty("java.specification.version");
+                }
+            });
+            return majorVersion(javaSpecVersion);
+        } catch (SecurityException e) {
```

Comment:
> if we think these types of errors should prevent the application from starting then that is fine.

Absolutely they should; again, the risks are either leaving the JVM in a questionable state or hiding bugs. No one benefits from that.

> I have also found it difficult in the past to catch exceptions thrown from a static context (this may have been bcz it was from Native code or for other reasons like stack trace was not populated) ... this makes the context an application developer can provide limited and makes locating these issues difficult.

Uncaught exception handlers can help here.